From fa82c6268754f0f1eed9f52ca8edfdebc7b74f13 Mon Sep 17 00:00:00 2001
From: "Daniel P. Berrange" <berrange@redhat.com>
Date: Fri, 16 Nov 2012 14:10:48 +0000
Subject: [PATCH] Don't set all typed parameters at once

It must not be assumed that all typed parameters which are read,
can also be written. Thus when setting typed parameters, the
list must be filtered down to only include the parameters that
the application has requested to change.

Signed-off-by: Daniel P. Berrange <berrange@redhat.com>
(cherry picked from commit b067d9beac1d65892e41aa81e1b52af70ac1012d)
---
 Virt.xs | 36 +++++++++++++++++++++++++-----------
 1 file changed, 25 insertions(+), 11 deletions(-)

diff --git a/Virt.xs b/Virt.xs
index 4795e56..76834f8 100644
--- a/Virt.xs
+++ b/Virt.xs
@@ -1248,18 +1248,30 @@ vir_typed_param_to_hv(virTypedParameter *params, int nparams)
 }
 
 
-static void
+static int
 vir_typed_param_from_hv(HV *newparams, virTypedParameter *params, int nparams)
 {
     unsigned int i;
     char * ptr;
     STRLEN len;
+    int ret = 0;
 
+    /* We only want to set parameters which we're actually changing
+     * so here we figure out which elements of 'params' we need to
+     * update, and overwrite the others
+     */
     for (i = 0 ; i < nparams ; i++) {
-        SV **val;
-
-        if (!hv_exists(newparams, params[i].field, strlen(params[i].field)))
+        if (!hv_exists(newparams, params[i].field, strlen(params[i].field))) {
+            if ((nparams-i) > 1)
+                memmove(params+i, params+i+1, sizeof(*params)*(nparams-(i+1)));
             continue;
+        }
+
+        ret++;
+    }
+
+    for (i = 0 ; i < ret ; i++) {
+        SV **val;
 
         val = hv_fetch (newparams, params[i].field, strlen(params[i].field), 0);
 
@@ -1294,6 +1306,8 @@ vir_typed_param_from_hv(HV *newparams, virTypedParameter *params, int nparams)
             break;
         }
     }
+
+    return ret;
 }
 
 
@@ -1705,7 +1719,7 @@ set_node_memory_parameters(conn, newparams, flags=0)
           _croak_error();
       }
 
-      vir_typed_param_from_hv(newparams, params, nparams);
+      nparams = vir_typed_param_from_hv(newparams, params, nparams);
 
       if (virNodeSetMemoryParameters(conn, params, nparams, flags) < 0)
           _croak_error();
@@ -3137,7 +3151,7 @@ set_scheduler_parameters(dom, newparams, flags=0)
               _croak_error();
           }
       }
-      vir_typed_param_from_hv(newparams, params, nparams);
+      nparams = vir_typed_param_from_hv(newparams, params, nparams);
       if (flags) {
           if (virDomainSetSchedulerParametersFlags(dom, params, nparams, flags) < 0)
               _croak_error();
@@ -3193,7 +3207,7 @@ set_memory_parameters(dom, newparams, flags=0)
           _croak_error();
       }
 
-      vir_typed_param_from_hv(newparams, params, nparams);
+      nparams = vir_typed_param_from_hv(newparams, params, nparams);
 
       if (virDomainSetMemoryParameters(dom, params, nparams, flags) < 0)
           _croak_error();
@@ -3245,7 +3259,7 @@ set_numa_parameters(dom, newparams, flags=0)
           _croak_error();
       }
 
-      vir_typed_param_from_hv(newparams, params, nparams);
+      nparams = vir_typed_param_from_hv(newparams, params, nparams);
 
       if (virDomainSetNumaParameters(dom, params, nparams, flags) < 0)
           _croak_error();
@@ -3297,7 +3311,7 @@ set_blkio_parameters(dom, newparams, flags=0)
           _croak_error();
       }
 
-      vir_typed_param_from_hv(newparams, params, nparams);
+      nparams = vir_typed_param_from_hv(newparams, params, nparams);
 
       if (virDomainSetBlkioParameters(dom, params, nparams,
                                       flags) < 0)
@@ -3738,7 +3752,7 @@ set_block_iotune(dom, disk, newparams, flags=0)
           _croak_error();
       }
 
-      vir_typed_param_from_hv(newparams, params, nparams);
+      nparams = vir_typed_param_from_hv(newparams, params, nparams);
       if (virDomainSetBlockIoTune(dom, disk, params, nparams, flags) < 0)
           _croak_error();
 
@@ -3789,7 +3803,7 @@ set_interface_parameters(dom, intf, newparams, flags=0)
           _croak_error();
       }
 
-      vir_typed_param_from_hv(newparams, params, nparams);
+      nparams = vir_typed_param_from_hv(newparams, params, nparams);
       if (virDomainSetInterfaceParameters(dom, intf, params, nparams, flags) < 0)
           _croak_error();
 
